contracts/03-tamagotchi-nft/src/lib.rs

```diff
@@ -1,19 +1,178 @@
+            if owner_transfered {
+                msg::reply(TmgEvent::Transferred(actor_id), 0)
+                    .expect("Error in sending reply");
+            }
+        },
+        TmgAction::Approve(actor_id) => {
+            let source_id = msg::source();
+            if tamagotchi.owner == source_id {
+                tamagotchi.approved_account = Some(actor_id);
+                msg::reply(TmgEvent::Approved(actor_id), 0)
+                    .expect("Error in sending reply");
+            }
+        },
+        TmgAction::RevokeApproval => {
+            let source_id = msg::source();
+            if tamagotchi.owner == source_id {
+                tamagotchi.approved_account = None;
+                msg::reply(TmgEvent::ApprovalRevoked, 0)
+                    .expect("Error in sending reply");
+            }
+        }
+    }
 }
 
 #[no_mangle]
 extern fn state() {
     // TODO: 0️⃣ Copy the `handle` function from the previous lesson and push changes to the master branch
+    msg::reply(state_ref(), 0)
+        .expect("Failed to share state");
+}
+
+
+fn blocks_height() -> u64 {
+    exec::block_height() as u64
+}
+
+fn state_ref() -> &'static Tamagotchi {
+    let state = unsafe { TAMAGOTCHI.as_ref() };
+    debug_assert!(state.is_some(), "State is not initialized");
+    unsafe { state.unwrap_unchecked() }
 }
+
+fn state_mut() -> &'static mut Tamagotchi {
+    let state = unsafe { TAMAGOTCHI.as_mut() };
+    debug_assert!(state.is_some(), "State is not initialized");
+    unsafe { state.unwrap_unchecked() } 
+}
+
+fn update_tamagotchi_fields() {
+    let state = state_mut();
+    let blocks_height = blocks_height();
+    state.fed = updated_field_value(
+        state.fed,
+        state.fed_block,
+        HUNGER_PER_BLOCK,
+        blocks_height
+    );
+    state.fed_block = blocks_height;
+    
+    state.entertained = updated_field_value(
+        state.entertained,
+        state.entertained_block,
+        BOREDOM_PER_BLOCK,
+        blocks_height
+    );
+    state.entertained_block = blocks_height;  
+      
+    state.rested = updated_field_value(
+        state.rested,
+        state.rested_block,
+        ENERGY_PER_BLOCK,
+        blocks_height
+    );
+    state.rested_block = blocks_height;    
+}
+
+fn updated_field_value(field: u64, field_block: u64, value_per_block: u64, blocks_height: u64) -> u64 {
+    let total_value_to_rest = (blocks_height - field_block) * value_per_block;
+    if field >= total_value_to_rest {
+        // If the given value of the tamagotchi is greater than the value to be 
+        // subtracted after a certain number of blocks, the update value is
+        // returned
+        field - total_value_to_rest
+    } else {
+        // If not, the given value is smaller, causing a negative result, zero
+        // is returned instead. 
+        0
+    }
+}
+
+fn update_field(field: &mut u64, increase_value: u64) {
+    *field += increase_value;
+    *field = *field.min(&mut 10_000);
+} 
```

Comment:
This function is not too idiomatic. It is better to return `Copy` types instead of mutating them.

```suggestion
fn update_field(field: u64, increase_value: u64) -> u64 {
    let field = field + increase_value;
    field.min(10_000)
}

```